🌱 Add conditions to the KubeadmControlPlane object

fabriziopandini opened the pull request:
**What this PR does / why we need it**:
This PR adds conditions to the KubeadmControlPlane object

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3137

fabriziopandini commented:
This PR requires https://github.com/kubernetes-sigs/cluster-api/pull/3136 to merge before

fabriziopandini commented on controlplane/kubeadm/api/v1alpha3/condition_consts.go:
```diff
@@ -0,0 +1,60 @@
+	// InitializedCondition documents that the control plane was initialized and the first control plane instance
+	// has completed the kubeadm init operation; as a consequence, there is an API server instance ready for
+	// processing requests.
+	InitializedCondition clusterv1.ConditionType = "Initialized"
```
@vincepri I have added this condition as agreed, but I think that what we really need is intead is control plane ready (name of the condition TBD)

vincepri replied:
We probably need both? Ready and Initialized

detiber replied:
I think it depends on what we want things to mean semantically.

Initialized has value today in that we are triggering on it as part of the contract for backwards compatibility reasons. 

From an external point of view, I think users really only care about Availability and if it is quiescent (how the CAEP defines Ready).

I think `Initialized` could be valuable in exposing progression towards quiescence.

fabriziopandini replied:
I'm -1 for both.
IMO the users are interested in knowing when the API server is available e.g. to apply addons, but I don't think there is value in reporting when the kubeadm config is initalized

vincepri replied:
Today, the Cluster object has both `ControlPlaneInitialized` and `ControlPlaneReady` without a clear definition of what those mean, let's take this opportunity to better define these based on what a user needs to see, and our controllers.

Initialized has been used often in controllers to understand if we can try to connect to the control plane after the first api-server machine is up. I do agree that maybe we're looking to express this a little better, for example:

- `Available`: The control plane is available to receive requests.
- `Ready`: The control plane is ready, every node in the control plane is also ready (needs CNI).

Initialized can be replaced by available, and we can gradually migrate controllers in the future to rely on that condition instead.

fabriziopandini replied:
Ok for Available

detiber commented on controlplane/kubeadm/controllers/controller.go:
```diff
@@ -222,6 +232,9 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 		return ctrl.Result{}, err
 	}
+	if certificates.HasGenerated() {
+		conditions.MarkTrue(kcp, controlplanev1.CertificatesGeneratedCondition)
```
Would external consumers actually care that certificates where generated or that certificates are available (or both)?

If watching externally for progress (or lack of progress), I'm wondering if it would make sense to possibly track both, rather than just if the certificates are generated.

fabriziopandini replied:
I will change into Available

detiber commented on controlplane/kubeadm/controllers/controller.go:
```diff
@@ -255,11 +268,23 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
+
+	// Aggregate the operational state of all the machines; while aggregating we are adding the
+	// source ref to the aggregate reason (reason@machine/name) so the problem can be easily tracked down to its source.
+	conditions.SetAggregateCondition(controlPlane.KCP, controlplanev1.MachinesReadyCondition, ownedMachines.ConditionGetters(), conditions.AddSourceRef())
```
From an external consumer point of view, would it make sense to try and differentiate between degraded and broken here somehow? Currently we leave that up to the user, but it seems like we should have enough context within KCP to differentiate between the two (especially since any upgrade or scaling operations would cause this to become False)

fabriziopandini replied:
I think this makes sense, but IMO those are separated Conditions

detiber commented on controlplane/kubeadm/controllers/controller.go:
```diff
@@ -255,11 +268,23 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
+		conditions.MarkFalse(controlPlane.KCP, controlplanev1.ControlPlaneUpgradingCondition, controlplanev1.UpgradingReason, clusterv1.ConditionSeverityWarning, "Upgrading to %s version (%d new version, %d old)", kcp.Spec.Version, kcp.Status.UpdatedReplicas, len(requireUpgrade))
 		return r.upgradeControlPlane(ctx, cluster, kcp, controlPlane)
+	default:
+		// make sure last upgrade operation is marked as completed (if any)
+		if conditions.Has(controlPlane.KCP, controlplanev1.ControlPlaneUpgradingCondition) {
+			conditions.MarkTrue(controlPlane.KCP, controlplanev1.ControlPlaneUpgradingCondition)
+		}
```
This feels a bit wrong. It doesn't seem like this is detecting success of upgrade here, but rather blindly setting the condition to true if present if we are not attempting an upgrade.

For example, if someone bumped the kubernetes version on their KCP, the new instance failed to come up correctly and they then reverted the change and deleted the broken new instance, this appears that it would mark the condition as if it were a successful upgrade.

Would it be better to track something like `HasRequestedConfig` (or possibly a better name) as the condition instead? That way the current logic would match the behavioral semantics of how we are setting the condition.

vincepri replied:
Makes sense, I got a little lost while reading the 3rd paragraph, but the problem brought up here can indeed happen. Without storing the wanted version, and the fact that the user has requested an upgrade, it'd be hard to reconcile this condition 🤔

vincepri replied:
_(not for today, otherwise we'll never finish this work)_ The pattern I've seen so far is that we want to set some conditions based by inspecting or performing operations, I was thinking in the future we could associate a function to each condition that performs an action when we need to evaluate it.

fabriziopandini replied:
As agreed, we are removing from this PR the ControlPlaneUpgradingCondition

detiber commented on controlplane/kubeadm/controllers/controller.go:
```diff
@@ -271,16 +296,24 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	case numMachines < desiredReplicas && numMachines == 0:
 		// Create new Machine w/ init
 		logger.Info("Initializing control plane", "Desired", desiredReplicas, "Existing", numMachines)
+		conditions.MarkFalse(controlPlane.KCP, controlplanev1.ResizeSucceededCondition, controlplanev1.ScalingUpReason, clusterv1.ConditionSeverityWarning, "Scaling up to %d replicas (actual %d)", desiredReplicas, numMachines)
```
Should this also set the `controlplanev1.InitializedCondition` to `False` here as well?

detiber commented on controlplane/kubeadm/controllers/controller.go:
```diff
@@ -271,16 +296,24 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 		logger.Info("Scaling down control plane", "Desired", desiredReplicas, "Existing", numMachines)
+		conditions.MarkFalse(controlPlane.KCP, controlplanev1.ResizeSucceededCondition, controlplanev1.ScalingDownReason, clusterv1.ConditionSeverityWarning, "Scaling down to %d replicas (actual %d)", desiredReplicas, numMachines)
 		return r.scaleDownControlPlane(ctx, cluster, kcp, controlPlane)
+	default:
+		// make sure last resize operation is marked as completed (if any)
+		if conditions.Has(controlPlane.KCP, controlplanev1.ResizeSucceededCondition) {
+			conditions.MarkTrue(controlPlane.KCP, controlplanev1.ResizeSucceededCondition)
+		}
```
Similar to above around upgrades, this doesn't necessarily feel like it's marking a resizing operation as successful as much as indicating the lack of resizing needed currently. 

I'm wondering if it might be better to name this condition so that it reflects that better. `HasRequestedSize` is the best I'm currently able to come up with, but I'm not overly happy with it either.

fabriziopandini replied:
As agreed, we are removing from this PR the ResizeSucceededCondition

detiber commented on controlplane/kubeadm/internal/etcd/util/set.go:
```diff
@@ -148,7 +148,7 @@ func (s UInt64Set) IsSuperset(s2 UInt64Set) bool {
 }
 
-// Equal returns true if and only if s1 is equal (as a set) to s2.
+// Equals returns true if and only if s1 is equal (as a set) to s2.
```
This looks like an accidental change
```suggestion
// Equal returns true if and only if s1 is equal (as a set) to s2.
```

detiber commented on controlplane/kubeadm/internal/machine_collection.go:
```diff
@@ -134,3 +135,13 @@ func (s FilterableMachineCollection) DeepCopy() FilterableMachineCollection {
 	return result
 }
+
+// ConditionGetters returns the slice with machines converted into conditions.Getter.
+func (s FilterableMachineCollection) ConditionGetters() []conditions.Getter {
+	res := make([]conditions.Getter, 0, len(s))
+	for _, v := range s {
+		value := *v
+		res = append(res, &value)
+	}
+	return res
+}
```
Once the conditions PR for Machines is merged, it looks like you could replace the call here by making `unsortedList` public and calling that instead.

fabriziopandini replied:
Unfortunately, this does not work: 

> cannot use ownedMachines.UnsortedList() (type []*"sigs.k8s.io/cluster-api/api/v1alpha3".Machine) as type []conditions.Getter in argument to conditions.SetAggregateCondition

fabriziopandini commented:
Removed WIP and rebased
As per discussion, only an initial set of conditions in left in this PR: MachinesReady, CertificatesAvailable, (control plane)Available 
@vincepri @detiber PTAL

vincepri commented:
/lgtm

/hold
for @detiber

vincepri commented:
/milestone v0.3.7

detiber commented on controlplane/kubeadm/controllers/status.go:
```diff
@@ -69,6 +70,7 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
 	// This only gets initialized once and does not change if the kubeadm config map goes away.
 	if status.HasKubeadmConfig {
 		kcp.Status.Initialized = true
+		conditions.MarkTrue(kcp, controlplanev1.AvailableCondition)
```
I'm not sure we should tie `AvailableCondition` to whether there is a kubeadmconfig. Should this check the `/healthz/ready` endpoint for the apiserver instead?

vincepri replied:
Both? :)

detiber replied:
Both definitely doesn't hurt :)

vincepri commented on controlplane/kubeadm/controllers/controller.go:
```diff
@@ -255,6 +267,11 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
+
 	requireUpgrade := controlPlane.MachinesNeedingUpgrade()
 	// Upgrade takes precedence over other operations
 	if len(requireUpgrade) > 0 {
```
```suggestion
	if len(needRollout) > 0 {
```

Can we change the name of these methods and variables to be `rollout` and not `upgrade`?

fabriziopandini replied:
done

fabriziopandini commented:
@vince, @detiber I have added more conditions documenting scaling up/down and rolling update.
Thank you very much for all the support and all the valuable tips that helped this PR to shape out

vince commented:
Wrong vince, @fabriziopandini . I think I get keep getting summoned here and I'd really prefer you stop it.

detiber commented:
@fabriziopandini linter doesn't appear happy, otherwise lgtm, especially since I can't think of better naming for the Resized condition.

detiber commented:
/test pull-cluster-api-test

detiber commented:
/lgtm

fabriziopandini commented:
@vincepri is it ok to unhold now?

detiber commented:
/lgtm

vincepri commented:
Unholding in a bit, wanted to get #3210 merged first, which should reduce the overall test running time

vincepri commented:
/hold cancel

vincepri commented:
/test pull-cluster-api-test
